File: py-src/fusion/storage/repository.py
from __future__ import annotations

from fusion import get_logger
from fusion.libs.model import Entity
from fusion.storage.commit import Commit
from fusion.storage.commit_graph import BranchMetadata, CommitGraph
from fusion.storage.delta import Delta, squash_deltas
from fusion.storage.hash_tree import (
    HangingSubtreesError,
    HashTree,
    build_hash_tree,
    update_hash_tree,
)
from fusion.storage.in_memory_store import InMemoryStore
from fusion.storage.in_memory_vcs_adapter import InMemoryVcsAdapter
from fusion.storage.vcs_adapter import InternalRepoUpdate, VcsAdapter
from fusion.util import get_new_id

log = get_logger(__name__)


class RepositoryIntegrityError(Exception):
    pass


class MissingBranchError(Exception):
    pass


class Repository:
    """Python port of the TS Repository.

    Always uses caching (head store + hash tree). The VcsAdapter is the
    persistence backend (InMemoryVcsAdapter by default).
    """

    def __init__(
        self,
        vcs_adapter: VcsAdapter | None = None,
        branch_name: str = "main",
    ) -> None:
        self._vcs_adapter: VcsAdapter = vcs_adapter or InMemoryVcsAdapter()
        self._current_branch = branch_name

        self._head_store = InMemoryStore()
        self._commit_graph = CommitGraph()
        self._commit_by_id: dict[str, Commit] = {}
        self._hash_tree: HashTree | None = None

    # ------------------------------------------------------------------
    # Factory methods
    # ------------------------------------------------------------------

    @staticmethod
    def create(
        vcs_adapter: VcsAdapter | None = None,
        branch_name: str = "main",
    ) -> Repository:
        """Create a fresh repository with an empty branch."""
        repo = Repository(vcs_adapter, branch_name)
        repo._commit_graph.create_branch(branch_name)
        repo._hash_tree = build_hash_tree(repo._head_store)

        # Persist initial branch
        repo._vcs_adapter.apply_update(
            InternalRepoUpdate(
                added_commits=[],
                removed_commits=[],
                updated_commits=[],
                added_branches=[BranchMetadata(name=branch_name, head_commit_id=None)],
                updated_branches=[],
                removed_branches=[],
            )
        )
        return repo

    @staticmethod
    def open(
        vcs_adapter: VcsAdapter,
        branch_name: str = "main",
        head_store_data: list[Entity] | None = None,
    ) -> Repository:
        """Open an existing repository, hydrating from the VcsAdapter."""
        repo = Repository(vcs_adapter, branch_name)

        commit_graph = repo._vcs_adapter.get_commit_graph()
        branch = commit_graph.branch(branch_name)
        if branch is None:
            all_branches = commit_graph.branches()
            raise MissingBranchError(
                f'Branch "{branch_name}" not found. '
                f"Branches: {[b.name for b in all_branches]}"
            )

        if head_store_data:
            for entity in head_store_data:
                repo._head_store.insert_one(entity)
            repo._hash_tree = build_hash_tree(repo._head_store)
        else:
            repo._hash_tree = build_hash_tree(repo._head_store)
            repo._hydrate_from_adapter()

        return repo

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    @property
    def head_store(self) -> InMemoryStore:
        return self._head_store

    @property
    def hash_tree(self) -> HashTree:
        if self._hash_tree is None:
            raise RuntimeError("Hash tree is not initialized")
        return self._hash_tree

    @property
    def current_branch(self) -> str:
        return self._current_branch

    def get_commit_graph(self) -> CommitGraph:
        return CommitGraph.from_data(self._commit_graph.data())

    def get_commits(self, ids: list[str]) -> list[Commit]:
        result: list[Commit] = []
        for commit_id in ids:
            commit = self._commit_by_id.get(commit_id)
            if commit is None:
                raise ValueError(f"Commit {commit_id} not found in cache")
            result.append(commit)
        return result

    def commit(
        self,
        delta: Delta,
        message: str = "",
        timestamp: float | None = None,
    ) -> Commit:
        """Create a commit from a delta.

        The delta is applied to the head store and a Commit object is produced
        and persisted through the VcsAdapter.

        Args:
            delta: The delta to commit.
            message: Commit message.
            timestamp: Epoch milliseconds. If None, uses current time.
        """
        import time

        if timestamp is None:
            timestamp = time.time() * 1000

        # Apply to the head store
        applied_delta = self._head_store.apply_delta(delta)

        # Update hash tree
        try:
            update_hash_tree(self.hash_tree, self._head_store, applied_delta)
        except HangingSubtreesError as e:
            # Revert store state
            self._head_store.apply_delta(applied_delta.reversed())
            raise RuntimeError(
                "Error updating hash tree. Are you committing entities whose "
                f"parents are not in the store? {e}"
            ) from e
        except Exception as e:
            self._head_store.apply_delta(applied_delta.reversed())
            raise RuntimeError(f"Error updating hash tree: {e}") from e

        snapshot_hash = self.hash_tree.root_hash()

        # Determine parent
        commits = self._commit_graph.branch_commits(self._current_branch)
        parent_id = commits[-1].id if commits else ""

        commit = Commit(
            id=get_new_id(),
            parent_id=parent_id,
            snapshot_hash=snapshot_hash,
            delta_data=applied_delta.asdict(),
            message=message,
            timestamp=timestamp,
        )

        # Update graph + cache
        self._commit_graph.add_commit(commit.metadata())
        self._commit_by_id[commit.id] = commit
        self._commit_graph.set_branch(self._current_branch, commit.id)

        branch = self._commit_graph.branch(self._current_branch)
        assert branch is not None

        self._vcs_adapter.apply_update(
            InternalRepoUpdate(
                added_commits=[commit],
                removed_commits=[],
                updated_commits=[],
                added_branches=[],
                updated_branches=[branch],
                removed_branches=[],
            )
        )

        return commit

    def reset(self, relative_to_head: int) -> None:
        """Reset the current branch back by *relative_to_head* commits (negative number)."""
        if relative_to_head == 0:
            return
        if relative_to_head > 0:
            raise ValueError("Reset forward not supported")

        head_commit = self._commit_graph.head_commit(self._current_branch)
        if head_commit is None:
            raise ValueError("No head commit found")

        branch_commits = self._commit_graph.branch_commits(self._current_branch)
        head_index = next(
            i for i, c in enumerate(branch_commits) if c.id == head_commit.id
        )

        target_index = head_index + relative_to_head
        if target_index < 0:
            raise ValueError("Resetting too far back")

        target_commit = branch_commits[target_index]

        # Revert by applying reversed deltas of commits being removed
        commits_to_revert = branch_commits[target_index + 1 :]
        commits_to_revert_full = self.get_commits([c.id for c in commits_to_revert])
        reversed_deltas = [
            Delta.from_data(c.delta_data).reversed().asdict()
            for c in commits_to_revert_full
        ]
        squashed_delta = squash_deltas(reversed_deltas)
        self._head_store.apply_delta(squashed_delta)

        # Remove from graph + cache
        for c in commits_to_revert:
            self._commit_by_id.pop(c.id, None)
            self._commit_graph.remove_commit(c.id)

        # Update hash tree
        update_hash_tree(self.hash_tree, self._head_store, squashed_delta)
        self._commit_graph.set_branch(self._current_branch, target_commit.id)

        # Integrity check
        snapshot_hash = self.hash_tree.root_hash()
        if snapshot_hash != target_commit.snapshot_hash:
            raise RepositoryIntegrityError("Snapshot hash mismatch after reset")

        branch = self._commit_graph.branch(self._current_branch)
        assert branch is not None
        self._vcs_adapter.apply_update(
            InternalRepoUpdate(
                added_commits=[],
                removed_commits=commits_to_revert,
                updated_commits=[],
                added_branches=[],
                updated_branches=[branch],
                removed_branches=[],
            )
        )

    def close(self) -> None:
        self._vcs_adapter.close()

    def erase_storage(self) -> None:
        self._vcs_adapter.erase_storage()

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _hydrate_from_adapter(self) -> None:
        """Replay all commits from the VcsAdapter onto the head store."""
        remote_graph = self._vcs_adapter.get_commit_graph()

        # Get branch commits in order
        branch = remote_graph.branch(self._current_branch)
        if branch is None or branch.head_commit_id is None:
            # Empty branch, nothing to hydrate
            self._commit_graph = remote_graph
            return

        branch_commits_meta = remote_graph.branch_commits(self._current_branch)
        commit_ids = [c.id for c in branch_commits_meta]
        commits = self._vcs_adapter.get_commits(commit_ids)

        # Replay deltas in order
        for commit_obj in commits:
            delta = Delta.from_data(commit_obj.delta_data)
            self._head_store.apply_delta(delta)
            self._commit_by_id[commit_obj.id] = commit_obj

        # Update hash tree for the final state
        self._hash_tree = build_hash_tree(self._head_store)

        # Verify integrity
        last_commit = commits[-1] if commits else None
        if last_commit and last_commit.snapshot_hash:
            snapshot_hash = self._hash_tree.root_hash()
            if snapshot_hash != last_commit.snapshot_hash:
                log.error(
                    "Snapshot hash mismatch: computed=%s, stored=%s, "
                    "commit_id=%s, entity_count=%d",
                    snapshot_hash,
                    last_commit.snapshot_hash,
                    last_commit.id,
                    len(list(self._head_store.find())),
                )
                raise RepositoryIntegrityError(
                    "Snapshot hash mismatch after hydration from VcsAdapter"
                )

        self._commit_graph = remote_graph

File: js-src/src/storage/repository/Repository.ts
import { getLogger } from "../../logging";
import { Commit, CommitData } from "../version-control/Commit";
import { CommitGraph, CommitGraphData } from "../version-control/CommitGraph";
import { VcsAdapter, InternalRepoUpdate } from "./VcsAdapter";
import { inferRepoChangesFromGraphs, sanityCheckAndHydrateInternalRepoUpdate } from "../management/sync-utils";
import { Delta, DeltaData, squashDeltas } from "../../model/Delta";
import { createId } from "../../util/base";
import { HangingSubtreesError, HashTree, buildHashTree, updateHashTree } from "../version-control/HashTree";
import { InMemoryStore, DEFAULT_INDEX_CONFIGS_LIST } from "../domain-store/InMemoryStore";
import { Entity, EntityData } from "../../model/Entity";
import { InMemoryVcsAdapter } from "./InMemoryVcsAdapter";
import { IndexedDBVcsAdapter } from "./IndexedDBVcsAdapter";
import { RestApiVcsAdapter } from "./RestApiVcsAdapter";
import { RestApiAuthConfig } from "../rest-api/Auth";
let log = getLogger('Repository')

export type VcsAdapterNames = "InMemory" | "IndexedDB" | "RestApi" | "InMemorySingletonForTesting";

// For running tests only
let _inmemadapter_instances_by_id: Map<string, InMemoryVcsAdapter> = new Map();

export function clearInMemoryAdapterInstances() {
    _inmemadapter_instances_by_id.clear();
}

export class RepositoryIntegrityError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "RepositoryIntegrityError";
    }
}

export class MissingBranchError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "MissingBranchError";
    }
}

export interface VcsAdapterArgs {
    projectId: string;
    localBranchName: string;
    baseUrl?: string;
    auth?: RestApiAuthConfig;
}

export interface VcsAdapterConfig {
    name: VcsAdapterNames
    args: VcsAdapterArgs;
}

export interface ResetFilter {
    relativeToHead: number;
}

export interface RepoUpdateData {
    commitGraph: CommitGraphData;
    upsertedCommits: CommitData[];
}

export interface CommitOptions {
    skipConflictingChanges?: boolean;
}

export async function getVcsAdapter(config: VcsAdapterConfig): Promise<VcsAdapter> {
    if (config.name === 'InMemory') {
        return new InMemoryVcsAdapter();
    } else if (config.name === 'IndexedDB') {
        const { projectId } = config.args;
        const adapter = new IndexedDBVcsAdapter(projectId);
        await adapter.initialize();
        return adapter;
    } else if (config.name === 'RestApi') {
        const { projectId, localBranchName, baseUrl, auth } = config.args;
        if (!baseUrl) {
            throw new Error("RestApi VCS adapter requires args.baseUrl in config");
        }
        if (!auth) {
            throw new Error("RestApi VCS adapter requires args.auth in config");
        }
        return new RestApiVcsAdapter(
            projectId,
            localBranchName,
            baseUrl,
            auth,
        );
    } else if (config.name === 'InMemorySingletonForTesting') {
        const { projectId } = config.args;
        if (!_inmemadapter_instances_by_id.has(projectId)) {
            const inMemAdapter = new InMemoryVcsAdapter();
            _inmemadapter_instances_by_id.set(projectId, inMemAdapter);
        }
        return _inmemadapter_instances_by_id.get(projectId)!;
    } else {
        throw new Error(`Unknown VCS adapter type: ${config.name}`);
    }
}

export class Repository {
    _vcsAdapter: VcsAdapter;
    private _adapterConfig: VcsAdapterConfig;
    private _isCaching: boolean;

    _currentBranch: string;
    private _headStore: InMemoryStore | null = null;
    _commitGraph: CommitGraph = new CommitGraph();  // Public only for testing purposes
    private _commitById: Map<string, Commit> = new Map();
    private _hashTree: HashTree | null = null;

    private constructor(
        storageAdapter: VcsAdapter,
        adapterConfig: VcsAdapterConfig,
        enableCaching: boolean) {

        this._vcsAdapter = storageAdapter;
        this._isCaching = enableCaching;
        this._adapterConfig = adapterConfig;
        this._currentBranch = adapterConfig.args.localBranchName;

        // Initialize the commit graph and head store
        if (enableCaching) {
            log.info(`Repository constructor: caching enabled. Using VCS adapter: ${adapterConfig.name}`);
            this._commitGraph = new CommitGraph();
            this._headStore = new InMemoryStore(DEFAULT_INDEX_CONFIGS_LIST);
        } else {
            log.info(`Repository constructor: caching disabled. Using VCS adapter: ${adapterConfig.name}`);
        }
    }

    private static async init(config: VcsAdapterConfig, enableCaching: boolean): Promise<Repository> {
        const storageAdapter = await getVcsAdapter(config);
        const repo = new Repository(storageAdapter, config, enableCaching);
        // Always init with defualt current branch. There's no repo without
        // a head. And then we'll pull from wherever
        repo._commitGraph.createBranch(repo._currentBranch);
        return repo
    }

    static async open(config: VcsAdapterConfig, enableCaching: boolean, headStoreData?: Entity<EntityData>[]): Promise<Repository> {
        let repo = await Repository.init(config, enableCaching);

        log.info(`Opening repository with caching: ${enableCaching}, VCS adapter: ${config.name}`);

        //
        const commitGraph = await repo._vcsAdapter.getCommitGraph();
        const allBranches = commitGraph.branches();
        const allCommits = commitGraph.commits();
        log.info(`[Repository.open] Loaded commit graph for branch "${repo._currentBranch}". ` +
            `Branches (${allBranches.length}): [${allBranches.map(b => b.name).join(', ')}], ` +
            `Commits: ${allCommits.length}`);

        let branch = commitGraph.branch(repo._currentBranch);
        if (branch === undefined) {
            throw new MissingBranchError(`Branch "${repo._currentBranch}" not found in the commit graph. ` +
                `Branches (${allBranches.length}): [${allBranches.map(b => b.name).join(', ')}], ` +
                `Commits: ${allCommits.length}, ` +
                `VCS adapter: ${config.name}, projectId: ${config.args.projectId}`);
        }

        if (!repo._isCaching) {
            log.info(`Repository opened without caching. Using VCS adapter: ${config.name}`);
            return repo;
        }

        if (headStoreData) {
            // Populate headStore from provided entity data (e.g. domain store)
            for (const entity of headStoreData) {
                repo.headStore.insertOne(entity);
            }
            log.info(`Populated headStore from provided data: ${headStoreData.length} entities`);
            repo._hashTree = await buildHashTree(repo.headStore);
        } else {
            // Initialize an empty hash tree before hydration so
            // _applyInternalUpdateToCache can update it incrementally.
            repo._hashTree = await buildHashTree(repo.headStore);
            await repo.hydrateCacheFromVcsAdapter();
        }

        log.info(`Repository opened with caching enabled. Using VCS adapter: ${config.name}`);
        return repo;
    }

    static async create(config: VcsAdapterConfig, enableCaching: boolean): Promise<Repository> {
        const repo = await Repository.init(config, enableCaching);

        // Create the default branch on the adappters
        // Locally the default is created in the constructor (to use for pull from adapetr)
        await repo._vcsAdapter.applyUpdate({
            addedCommits: [],
            removedCommits: [],
            updatedCommits: [],
            addedBranches: [{ name: repo._currentBranch, headCommitId: null }],
            updatedBranches: [],
            removedBranches: []
        });

        log.info(`Created repository with caching: ${enableCaching}, current branch: ${repo._currentBranch}`);
        if (repo._isCaching) {
            repo._hashTree = await buildHashTree(repo.headStore);
        }
        return repo;
    }

    get headStore(): InMemoryStore {
        if (!this._headStore) {
            throw new Error("Head store is not available. Caching might be disabled.");
        }
        return this._headStore;
    }

    get hashTree(): HashTree {
        if (!this._hashTree) {
            throw new Error("Hash tree is not initialized");
        }
        return this._hashTree;
    }

    async getCommitGraph(): Promise<CommitGraph> {
        if (this._isCaching) {
            // Return a defensive clone so callers cannot mutate internal graph
            return CommitGraph.fromData(this._commitGraph.data());
        }
        return await this._vcsAdapter.getCommitGraph();
    }

    async getCommits(ids: string[]): Promise<Commit[]> {
        if (this._isCaching) {
            const commits: Commit[] = [];
            for (const id of ids) {
                const commit = this._commitById.get(id);
                if (!commit) {
                    throw new Error(`Commit ${id} not found in cache`);
                }
                commits.push(commit);
            }
            return commits;
        }
        return this._vcsAdapter.getCommits(ids);
    }

    async commit(delta: Delta, message: string, options: CommitOptions = {}): Promise<Commit> {
        if (!this._currentBranch) {
            throw new Error("Current branch is not set");
        }
        if (!this._isCaching) {
            throw new Error("Cannot create a commit without head state caching enabled");
        }

        log.info('Committing', delta, message)
        // Apply to the head store
        const appliedDelta = this.headStore.applyDelta(delta, undefined, options.skipConflictingChanges === true);

        // Get snapshotHash
        try {
            await updateHashTree(this.hashTree, this.headStore, appliedDelta)
        } catch (e) {
            this.headStore.applyDelta(appliedDelta.reversed()); // Restore the store state, commit is unsuccessful
            if (e instanceof HangingSubtreesError) {
                throw Error('Error updating hashtree. You\'re probably trying to commit objects to the store who\'s parents are not present in it. Error: ' + e)
            }
            throw Error('Error updating hash tree: ' + e)
        }
        let snapshotHash = this.hashTree.rootHash()

        // Create commit
        const commitGraph = this._commitGraph!;
        let commits = commitGraph.branchCommits(this._currentBranch);

        let parentId = ''
        if (commits.length > 0) {
            // Initial commit
            parentId = commits.at(-1)!.id
        }
        let commit = new Commit({
            id: createId(),
            parentId: parentId,
            snapshotHash: snapshotHash,
            deltaData: appliedDelta.data,
            message: message,
            timestamp: Date.now()
        })

        // Add commit to sync graph
        commitGraph.addCommit(commit.metadata())
        this._commitById.set(commit.id, commit)
        commitGraph.setBranch(this._currentBranch, commit.id)

        const branch = commitGraph.branch(this._currentBranch)!;

        const internalUpdate: InternalRepoUpdate = {
            addedCommits: [commit],
            removedCommits: [],
            updatedCommits: [],
            addedBranches: [],
            updatedBranches: [branch],
            removedBranches: []
        };
        await this._vcsAdapter.applyUpdate(internalUpdate);

        return commit;
    }

    async createBranch(branchName: string): Promise<void> {
        if (!this._isCaching) {
            // In non-caching mode, this should probably be an atomic operation
            // on the storage adapter, which is not implemented.
            // For now, we only support this in caching mode.
            throw new Error("Cannot create a branch without head state caching enabled");
        }
        const commitGraph = this._commitGraph!;
        commitGraph.createBranch(branchName);
        const branch = commitGraph.branch(branchName)!;

        const update: InternalRepoUpdate = {
            addedCommits: [],
            removedCommits: [],
            updatedCommits: [],
            addedBranches: [branch],
            updatedBranches: [],
            removedBranches: []
        };
        await this._vcsAdapter.applyUpdate(update);
        // throw Error(`Created branch ${JSON.stringify(await this._storageAdapter.getCommitGraph())} for name ${branchName}`)
    }

    async pull(repository: Repository | VcsAdapter) {
        let remoteGraph = await repository.getCommitGraph()
        const ownGraph = await this.getCommitGraph();

        let repoUpdateSlim = inferRepoChangesFromGraphs(ownGraph, remoteGraph);

        // Get upserted commits in full (added + updated)
        const upsertIds = [
            ...repoUpdateSlim.addedCommits.map((c) => c.id),
            ...repoUpdateSlim.updatedCommits.map((c) => c.id)
        ];
        let upsertedCommits = upsertIds.length > 0 ? await repository.getCommits(upsertIds) : [];

        // Sanity check and hydrate (added + updated) with the upserted commits (with deltas)
        let repoUpdate = sanityCheckAndHydrateInternalRepoUpdate(repoUpdateSlim, upsertedCommits);

        // Persist the changes to the underlying storage first
        await this._vcsAdapter.applyUpdate(repoUpdate);

        if (this._isCaching) {
            await this._applyInternalUpdateToCache(repoUpdate, remoteGraph);
        }
    }

    async hydrateCacheFromVcsAdapter() {
        if (!this._isCaching) {
            throw new Error("Cannot hydrate cache without caching enabled");
        }

        log.info('Hydrating cache from VCS adapter');
        // Get the commit graph from the storage adapter
        const remoteGraph = await this._vcsAdapter.getCommitGraph();
        const ownGraph = await this.getCommitGraph();

        let repoUpdateSlim = inferRepoChangesFromGraphs(ownGraph, remoteGraph);

        // Get upserted commits in full (added + updated)
        const upsertIds = [
            ...repoUpdateSlim.addedCommits.map((c) => c.id),
            ...repoUpdateSlim.updatedCommits.map((c) => c.id)
        ];
        let upsertedCommits = upsertIds.length > 0 ? await this._vcsAdapter.getCommits(upsertIds) : [];

        let repoUpdate = sanityCheckAndHydrateInternalRepoUpdate(repoUpdateSlim, upsertedCommits);

        await this._applyInternalUpdateToCache(repoUpdate, remoteGraph);
    }

    async applyRepoUpdate(updateInfo: RepoUpdateData): Promise<void> {
        let remoteGraph = CommitGraph.fromData(updateInfo.commitGraph);
        let upsertedCommits = updateInfo.upsertedCommits.map((data) => new Commit(data));

        // Form the internalRepoUpdate object
        let repoUpdateSlim = inferRepoChangesFromGraphs(this._commitGraph, remoteGraph);
        // Sanity check and hydrate addedCommits with the upsertedCommits (with deltas included)
        let repoUpdate = sanityCheckAndHydrateInternalRepoUpdate(repoUpdateSlim, upsertedCommits);

        // Persist the changes to the underlying storage first
        await this._vcsAdapter.applyUpdate(repoUpdate);

        if (this._isCaching) {
            await this._applyInternalUpdateToCache(repoUpdate, remoteGraph);
        }
    }

    async reset(filter: ResetFilter): Promise<void> {
        if (!this._currentBranch) {
            throw new Error("Current branch is not set");
        }
        if (!this._isCaching) {
            throw new Error("Cannot reset without head state caching enabled");
        }
        const commitGraph = this._commitGraph!;

        // Update the head store state to reflect the requested branch/head pos
        let { relativeToHead } = filter

        if (relativeToHead === 0) {
            return
        } else if (relativeToHead > 0) {
            throw new Error("Reset forward not supported")
        }

        let headCommit = commitGraph.headCommit(this._currentBranch)
        if (!headCommit) {
            throw new Error("No head commit found")
        }

        let branchCommits = commitGraph.branchCommits(this._currentBranch)
        let index = branchCommits.indexOf(headCommit)
        if (index === -1) {
            throw new Error("Head commit not found in branch commits")
        }

        let targetIndex = index + relativeToHead
        if (targetIndex < 0) {
            throw new Error("Resetting too far back")
        }

        let targetCommit = branchCommits.at(targetIndex)
        if (!targetCommit) {
            throw new Error("Target commit not found")
        }

        // Revert the head store state by applying the deltas for the removed
        // commits in reverse
        let commitsToRevert = branchCommits.slice(targetIndex + 1)
        let commitsToRevertFull = await this.getCommits(commitsToRevert.map(c => c.id));
        const reversedDeltas = commitsToRevertFull.map(c => new Delta(c.deltaData as DeltaData).reversed().data);
        const squashedDelta = squashDeltas(reversedDeltas);
        this.headStore.applyDelta(squashedDelta);


        // Remove from commit graph and local commits
        for (let commit of commitsToRevert) {
            this._commitById.delete(commit.id)
            commitGraph.removeCommit(commit.id)
        }

        // Update hash tree
        await updateHashTree(this.hashTree, this.headStore, squashedDelta)
        commitGraph.setBranch(this._currentBranch, targetCommit.id)

        // Assert that the hash is correct
        let snapshotHash = this.hashTree.rootHash()
        if (snapshotHash !== targetCommit.snapshotHash) {
            throw new RepositoryIntegrityError("Snapshot hash of the head store state does not match the one of the applied commit (on reset)")
        }

        // Persist changes
        const branch = commitGraph.branch(this._currentBranch)!;
        const update: InternalRepoUpdate = {
            addedCommits: [],
            removedCommits: commitsToRevert,
            updatedCommits: [],
            addedBranches: [],
            updatedBranches: [branch],
            removedBranches: []
        };
        await this._vcsAdapter.applyUpdate(update);
    }

    async _applyInternalUpdateToCache(repoChanges: InternalRepoUpdate, remoteGraph: CommitGraph): Promise<void> {
        // If caching, apply the now-persisted changes to the in-memory cache
        const cacheGraph = this._commitGraph!;

        let {
            addedCommits,
            removedCommits,
            updatedCommits,
            addedBranches,
            updatedBranches,
            removedBranches
        } = repoChanges

        // Do the commit removal
        removedCommits.forEach((commit) => {
            this._commitById.delete(commit.id)
            cacheGraph.removeCommit(commit.id)
        })

        // Update commits (full replace: metadata + delta)
        updatedCommits.forEach((commit) => {
            // Replace in map
            this._commitById.set(commit.id, commit)
            // Refresh metadata in graph
            cacheGraph.removeCommit(commit.id)
            cacheGraph.addCommit(commit.metadata())
        })

        // Add new commits to the local map for further processing
        addedCommits.forEach((commit) => {
            this._commitById.set(commit.id, commit)
            cacheGraph.addCommit(commit.metadata())
        })

        let commitsBehind: string[] = [] // ids
        let remoteHeadCommit = remoteGraph.headCommit(this._currentBranch!)
        let localHeadCommit = cacheGraph.headCommit(this._currentBranch!)

        if (remoteHeadCommit) {
            let remoteHeadId = remoteHeadCommit.id
            let localHeadId = localHeadCommit ? localHeadCommit.id : null
            if (remoteHeadId !== localHeadId) {
                // Get the commits behind
                let behind = remoteGraph.commitsBetween(localHeadId, remoteHeadId)
                commitsBehind = behind.map((c) => c.id)
            }
        } else {
            if (localHeadCommit) {
                throw new Error("Irrational changes - remote branch empty, while local is not")
            }
        }

        if (commitsBehind.length === 0) {
            log.info('No new commits to apply to headStore.')
        } else {
            // Squash deltas and apply the update to the head store
            let deltas: DeltaData[] = []

            for (let commitId of commitsBehind) {
                let commit = this._commitById.get(commitId)
                if (commit && commit.deltaData) {
                    deltas.push(commit.deltaData)
                } else {
                    throw new Error(`Critical: Missing commit or deltaData for ${commitId}`)
                }
            }

            // Apply to head store
            let squashedDelta = squashDeltas(deltas)
            this.headStore.applyDelta(squashedDelta)

            // Update the hash tree
            await updateHashTree(this.hashTree, this.headStore, squashedDelta)

            // Assert hash is correct
            let snapshotHash = this.hashTree.rootHash()
            if (snapshotHash !== remoteHeadCommit!.snapshotHash) {
                log.error('Snapshot hash mismatch after pull. Remote head commit:', remoteHeadCommit, 'Hash tree root:', this.hashTree.rootHash())
                throw new RepositoryIntegrityError("Snapshot hash mismatch after pull")
            }
        }

        // Apply branch changes to the cache
        addedBranches.forEach((branch) => {
            cacheGraph.createBranch(branch.name)
        })
        updatedBranches.forEach((branch) => {
            cacheGraph.setBranch(branch.name, branch.headCommitId)
        })
        removedBranches.forEach((branch) => {
            if (branch.name === this._currentBranch) {
                throw new Error("Cannot remove current branch")
            }
            cacheGraph.removeBranch(branch.name)
        })
    }

    async eraseStorage(): Promise<void> {
        return this._vcsAdapter.eraseStorage();
    }

    close() {
        log.info('Repo close called. Closing VCS adapter.');
        this._vcsAdapter.close();
    }
}


export  async function verifyRepositoryIntegrity(repository: Repository | VcsAdapter, branchName: string): Promise<boolean> {
    log.info(`Verifying integrity of repository for branch "${branchName}"`);

    const commitGraph: CommitGraph = await repository.getCommitGraph();
    const branchCommitsMinimal = commitGraph.branchCommits(branchName);

    if (branchCommitsMinimal.length === 0) {
        log.info("No commits in branch, integrity check skipped.");
        return true;
    }

    const commitIds = branchCommitsMinimal.map(c => c.id);
    const branchCommits = await repository.getCommits(commitIds);

    const store = new InMemoryStore();
    let hashTree: HashTree = await buildHashTree(store);

    for (const commit of branchCommits) {
        log.info(`Verifying commit ${commit.id}`);
        const delta = new Delta(commit.deltaData);

        // Apply delta to the in-memory store
        store.applyDelta(delta);

        // Update the hash tree
        await updateHashTree(hashTree, store, delta);

        // Compare the hash tree's root hash with the commit's snapshot hash
        const calculatedHash = hashTree.rootHash();
        if (calculatedHash !== commit.snapshotHash) {
            let deltaSize = Object.keys(commit.deltaData).length
            let deltaDataStr = '';
            if (deltaSize < 100) {
                deltaDataStr = JSON.stringify(commit.deltaData, null, 2);
            }
            log.error(`Integrity check failed at commit ${commit.id}: expected hash ${commit.snapshotHash}, but got ${calculatedHash}. Delta size is ${deltaSize}. Delta data:\n ${deltaDataStr}`);

            return false;
        }
    }

    log.info("Repository integrity verified successfully.");
    return true;
}
